File: src/IsbInterop/Accessory/Proxies/Application.vb
Imports IsbInterop.ComponentTokens
Imports IsbInterop.ComponentTokens.Proxies
Imports IsbInterop.Data
Imports IsbInterop.Data.Proxies
Imports IsbInterop.DataTypes.Enumerable
Imports IsbInterop.EDocuments
Imports IsbInterop.EDocuments.Proxies
Imports IsbInterop.Folders
Imports IsbInterop.Folders.Proxies
Imports IsbInterop.References
Imports IsbInterop.References.Proxies
Imports IsbInterop.Scripts
Imports IsbInterop.Scripts.Proxies
Imports IsbInterop.Searches
Imports IsbInterop.Searches.Proxies
Imports IsbInterop.ServiceForms
Imports IsbInterop.ServiceForms.Proxies
Imports IsbInterop.Tasks
Imports IsbInterop.Tasks.Proxies

Namespace Accessory.Proxies

  ''' <summary>
  ''' Обертка над IApplication.
  ''' </summary>
  Friend Class Application
    Inherits BaseIsbObject
    Implements IApplication

    #Region "Поля и свойства"

    ''' <summary>
    ''' ИД процесса Application.
    ''' </summary>
    Public ReadOnly Property PID As Integer Implements IApplication.PID
      Get
        Return GetRcwProperty("PID")
      End Get
    End Property

    #End Region

    #Region "Методы"

    ''' <summary>
    ''' Получает фабрику вариантов запуска.
    ''' </summary>
    ''' <returns>Фабрика вариантов запуска.</returns>
    Public Function GetComponentTokenFactory() As IComponentTokenFactory Implements IApplication.GetComponentTokenFactory
      Dim rcwIComponentTokenFactory = GetRcwProperty("ComponentTokenFactory")
      Return New ComponentTokenFactory(rcwIComponentTokenFactory, Scope)
    End Function

    ''' <summary>
    ''' Создает блокировку.
    ''' </summary>
    ''' <param name="objectType">Тип объекта.</param>
    ''' <param name="objectId">ИД объекта.</param>
    ''' <returns>Объект-блокировка.</returns>
    Public Function CreateLock(objectType As TCompType, objectId As Integer) As ILock Implements IApplication.CreateLock
      Dim rcwILock = InvokeRcwInstanceMethod("CreateLock", New Object() {objectType, objectId})
      Return New Lock(rcwILock, Scope)
    End Function

    ''' <summary>
    ''' Получает объект-соединение, посредством которого приложение связано с SQL-сервером.
    ''' </summary>
    ''' <returns>Объект-соединение.</returns>
    Public Function GetConnection() As IConnection Implements IApplication.GetConnection
      Dim rcwIConnection = GetRcwProperty("Connection")
      Return New Connection(rcwIConnection, Scope)
    End Function

    ''' <summary>
    ''' Получает фабрику электронных документов.
    ''' </summary>
    ''' <returns>Фабрика электронных документов.</returns>
    Public Function GetEDocumentFactory() As IEDocumentFactory Implements IApplication.GetEDocumentFactory
      Dim rcwIEDocumentFactory = GetRcwProperty("EDocumentFactory")
      Return New EDocumentFactory(rcwIEDocumentFactory, Scope)
    End Function

    ''' <summary>
    ''' Получает фабрику папок.
    ''' </summary>
    ''' <returns>Фабрика папок.</returns>
    Function GetFolderFactory() As IFolderFactory Implements IApplication.GetFolderFactory
      Dim rcwIFolderFactory = GetRcwProperty("FolderFactory")
      Return New FolderFactory(rcwIFolderFactory, Scope)
    End Function

    ''' <summary>
    ''' Получает фабрику заданий.
    ''' </summary>
    ''' <returns>Фабрика заданий</returns>
    Public Function GetJobFactory() As IJobFactory Implements IApplication.GetJobFactory
      Dim rcwIJobFactory = GetRcwProperty("JobFactory")
      Return New JobFactory(rcwIJobFactory, Scope)
    End Function

    ''' <summary>
    ''' Получает фабрику типов справочников.
    ''' </summary>
    ''' <returns>Фабрика типов справочников.</returns>
    Public Function GetReferencesFactory() As IReferencesFactory Implements IApplication.GetReferencesFactory
      Dim rcwIReferencesFactory = GetRcwProperty("ReferencesFactory")
      Return New ReferencesFactory(rcwIReferencesFactory, Scope)
    End Function

    ''' <summary>
    ''' Получает фабрику сценариев.
    ''' </summary>
    ''' <returns>Фабрика сценариев.</returns>
    Public Function GetScriptFactory() As IScriptFactory Implements IApplication.GetScriptFactory
      Dim rcwIScriptFactory = GetRcwProperty("ScriptFactory")
      Return New ScriptFactory(rcwIScriptFactory, Scope)
    End Function

    ''' <summary>
    ''' Получает фабрику поисков.
    ''' </summary>
    ''' <returns>Фабрика поисков.</returns>
    Public Function GetSearchFactory() As ISearchFactory Implements IApplication.GetSearchFactory
      Dim rcwISearchFactory = GetRcwProperty("SearchFactory")
      Return New SearchFactory(rcwISearchFactory, Scope)
    End Function

    ''' <summary>
    ''' Получает фабрику служебных объектов.
    ''' </summary>
    ''' <returns>Фабрика служебных объектов.</returns>
    Public Function GetServiceFactory() As IServiceFactory Implements IApplication.GetServiceFactory
      Dim rcwIServiceFactory = GetRcwProperty("ServiceFactory")
      Return New ServiceFactory(rcwIServiceFactory, Scope)
    End Function

    ''' <summary>
    ''' Получает фабрику задач.
    ''' </summary>
    ''' <returns>Фабрика задач.</returns>
    Public Function GetTaskFactory() As ITaskFactory Implements IApplication.GetTaskFactory
      Dim rcwITaskFactory = GetRcwProperty("TaskFactory")
      Return New TaskFactory(rcwITaskFactory, Scope)
    End Function

    ''' <summary>
    ''' Завершает работу с приложением.
    ''' </summary>
    Public Sub DoFinalize() Implements IApplication.DoFinalize
      InvokeRcwInstanceMethod("Finalize")
    End Sub

    #End Region

    #Region "Конструкторы"

    ''' <summary>
    ''' Конструктор.
    ''' </summary>
    ''' <param name="rcwApplication">COM-объект IApplication.</param>
    ''' <param name="scope">Область видимости.</param>
    Friend Sub New(rcwApplication As Object, scope As IScope)
      MyBase.New(rcwApplication, scope)
    End Sub

    #End Region

  End Class
End Namespace

File: src/IsbInterop/System/Proxies/Application.cs
using IsbInterop.Accessory;
using IsbInterop.Accessory.Proxies;
using IsbInterop.ComponentTokens;
using IsbInterop.ComponentTokens.Proxies;
using IsbInterop.Data;
using IsbInterop.Data.Proxies;
using IsbInterop.DataTypes.Enumerable;
using IsbInterop.EDocuments;
using IsbInterop.EDocuments.Proxies;
using IsbInterop.Folders;
using IsbInterop.Folders.Proxies;
using IsbInterop.References;
using IsbInterop.References.Proxies;
using IsbInterop.Scripts;
using IsbInterop.Scripts.Proxies;
using IsbInterop.Searches;
using IsbInterop.Searches.Proxies;
using IsbInterop.ServiceForms;
using IsbInterop.ServiceForms.Proxies;
using IsbInterop.Tasks;
using IsbInterop.Tasks.Proxies;

namespace IsbInterop.System.Proxies
{
  /// <summary>
  /// Обертка над IApplication.
  /// </summary>
  internal class Application : BaseIsbObject, IApplication
  {
    #region Поля и свойства

    /// <summary>
    /// ИД процесса.
    /// </summary>
    public int PID => (int)GetRcwProperty("PID");

    /// <summary>
    /// Контекст организации.
    /// </summary>
    public string OurFirmContext
    {
      get => (string)GetRcwProperty("OurFirmContext");
      set => SetRcwProperty("OurFirmContext", value);

    }

    #endregion

    #region Методы

    /// <summary>
    /// Получает фабрику вариантов запуска.
    /// </summary>
    /// <returns>Фабрика вариантов запуска.</returns>
    public IComponentTokenFactory GetComponentTokenFactory()
    {
      var rcwIComponentTokenFactory = GetRcwProperty("ComponentTokenFactory");
      return new ComponentTokenFactory(rcwIComponentTokenFactory, Scope);
    }

    /// <summary>
    /// Создает блокировку.
    /// </summary>
    /// <param name="objectType">Тип объекта.</param>
    /// <param name="objectId">ИД объекта.</param>
    /// <returns>Объект-блокировка.</returns>
    public ILock CreateLock(TCompType objectType, int objectId)
    {
      var rcwILock = InvokeRcwInstanceMethod("CreateLock", new object[] { objectType, objectId });
      return new Lock(rcwILock, Scope);
    }

    /// <summary>
    /// Получает объект-соединение.
    /// </summary>
    /// <returns>Объект-соединение.</returns>
    public IConnection GetConnection()
    {
      var rcwIConnection = GetRcwProperty("Connection");
      return new Connection(rcwIConnection, Scope);
    }

    /// <summary>
    /// Получает фабрику электронных документов.
    /// </summary>
    /// <returns>Фабрика электронных документов.</returns>
    public IEDocumentFactory GetEDocumentFactory()
    {
      var rcwIEDocumentFactory = GetRcwProperty("EDocumentFactory");
      return new EDocumentFactory(rcwIEDocumentFactory, Scope);
    }

    /// <summary>
    /// Получает фабрику папок.
    /// </summary>
    /// <returns>Фабрика папок.</returns>
    public IFolderFactory GetFolderFactory()
    {
      var rcwIFolderFactory = GetRcwProperty("FolderFactory");
      return new FolderFactory(rcwIFolderFactory, Scope);
    }

    /// <summary>
    /// Получает фабрику заданий.
    /// </summary>
    /// <returns>Фабрика заданий.</returns>
    public IJobFactory GetJobFactory()
    {
      var rcwIJobFactory = GetRcwProperty("JobFactory");
      return new JobFactory(rcwIJobFactory, Scope);
    }

    /// <summary>
    /// Получает фабрику типов справочников.
    /// </summary>
    /// <returns>Фабрика типов справочников.</returns>
    public IReferencesFactory GetReferencesFactory()
    {
      var rcwIReferencesFactory = GetRcwProperty("ReferencesFactory");
      return new ReferencesFactory(rcwIReferencesFactory, Scope);
    }

    /// <summary>
    /// Получает фабрику сценариев.
    /// </summary>
    /// <returns>Фабрика сценариев.</returns>
    public IScriptFactory GetScriptFactory()
    {
      var rcwIScriptFactory = GetRcwProperty("ScriptFactory");
      return new ScriptFactory(rcwIScriptFactory, Scope);
    }

    /// <summary>
    /// Получает фабрику поисков.
    /// </summary>
    /// <returns>Фабрика поисков.</returns>
    public ISearchFactory GetSearchFactory()
    {
      var rcwISearchFactory = GetRcwProperty("SearchFactory");
      return new SearchFactory(rcwISearchFactory, Scope);
    }

    /// <summary>
    /// Получает фабрику служебных объектов.
    /// </summary>
    /// <returns>Фабрика служебных объектов.</returns>
    public IServiceFactory GetServiceFactory()
    {
      var rcwIServiceFactory = GetRcwProperty("ServiceFactory");
      return new ServiceFactory(rcwIServiceFactory, Scope);
    }

    /// <summary>
    /// Получает фабрику задач.
    /// </summary>
    /// <returns>Фабрика задач.</returns>
    public ITaskFactory GetTaskFactory()
    {
      var rcwITaskFactory = GetRcwProperty("TaskFactory");
      return new TaskFactory(rcwITaskFactory, Scope);
    }

    /// <summary>
    /// Завершает работу с приложением.
    /// </summary>
    public void DoFinalize()
    {
      InvokeRcwInstanceMethod("Finalize");
    }

    #endregion

    #region Конструкторы

    /// <summary>
    /// Конструктор.
    /// </summary>
    /// <param name="rcwApplication">COM-объект IApplication.</param>
    /// <param name="scope">Область видимости.</param>
    internal Application(object rcwApplication, IScope scope) : base(rcwApplication, scope) { }

    #endregion
  }
}